SJRK-237: Reorganize project folder structure and theming

BlueSlug opened the pull request:
Substantial changes were made, please see [the Jira](https://issues.fluidproject.org/browse/SJRK-237) for more information.

Some tests were not added due to a race condition relating to a bug with UIO implementation. This has been captured in [SJRK-255](https://issues.fluidproject.org/browse/SJRK-255)

@cindyli to provide review :)

BlueSlug committed:
SJRK-237: added serverConfig example file, adjusted gitignore to match

BlueSlug committed:
SJRK-237: major reorganization of the folder structure. very broken.

BlueSlug committed:
SJRK-237: fixed incorrect quotes in serverConfig example file

BlueSlug committed:
SJRK-237: fixed path references in server test files

BlueSlug committed:
SJRK-237: updated paths and got static tests working again

BlueSlug committed:
Merge remote-tracking branch 'upstream/stories-floe-dev' into SJRK-237

BlueSlug committed:
Merge remote-tracking branch 'upstream/stories-floe-dev' into SJRK-237

BlueSlug committed:
Merge remote-tracking branch 'upstream/stories-floe-dev' into SJRK-237

BlueSlug committed:
SJRK-237: implemented top-level config file for server

BlueSlug committed:
SJRK-237: added request handlers for each theme

BlueSlug committed:
SJRK-237: Moved HTML files to root dirs of their respective themes

BlueSlug committed:
SJRK-237: set up fallback order for serving theme files (baseTheme last)

cindyli committed:
SJRK-237: Adjusted the order of request handlers in serverSetup to fix one issue with loading the home page.

BlueSlug committed:
SJRK-237: updated CSS and JS file paths to reflect new folder structure

BlueSlug committed:
SJRK-237: updated browser tests to reflect new route for UI JS files

BlueSlug committed:
SJRK-237: parameterized the custom theme to be determined by config file
The base theme pages are a little broken still and need to be updated to 
load the theme in programatically rather than having it hardcoded as it 
is currently.

BlueSlug committed:
SJRK-237: added a request handler and wiring to expose config values

BlueSlug committed:
SJRK-237: fixed issues with custom theme loading

BlueSlug committed:
SJRK-237: made some fixes to the Karisma theme to ensure it works enough

BlueSlug committed:
Merge remote-tracking branch 'upstream/stories-floe-dev' into SJRK-237

BlueSlug committed:
SJRK-237: Now serves base theme properly when no theme is specified

BlueSlug committed:
SJRK-237: added test files for storyTellingServerUI.js and prelim. tests

BlueSlug committed:
SJRK-237: refactored loadThemedPage() and added tests for it

BlueSlug committed:
SJRK-237: added ServerUi tests to all-tests, +bugfix and comments update

BlueSlug committed:
SJRK-237: fixed race condition and added more tests for loadThemedPage()

BlueSlug committed:
SJRK-237: Added test for CSS file loading

BlueSlug committed:
SJRK-237: added file instance count to CSS file test function

BlueSlug committed:
SJRK-237: relocated testutil functions to share them with upcoming tests

BlueSlug committed:
SJRK-237: promisified functions to avoid race conditions, added options

BlueSlug committed:
SJRK-237: added config and test file for custom theme and clientConfig

BlueSlug committed:
SJRK-237: removed some cruft

BlueSlug committed:
SJRK-237: fixed custom theme server test and removed extraneous images

BlueSlug committed:
SJRK-237: added server theme config fallback tests
- there are issues running them together as part of the all-tests 
battery, so they will be ommitted from that list for now.

BlueSlug committed:
SJRK-237: moved upload folder specifications back to serverSetup.js
This avoids accidental reconfiguration of these values by someone who 
just wants to customise the theme or turn saving on or off.

BlueSlug committed:
SJRK-237: updated docs and moved grade list to new file (to be updated)

BlueSlug committed:
SJRK-237: updated grades list documentation and added server grades

BlueSlug committed:
SJRK-237: added theme fallback testing to all-tests. Order is important.
The order of the tests is important due to a race condition being 
encountered while setting up and tearing down each test sequence. This 
order ensures that the base theme config doesn't override anything in 
subsequent sequences which use the learningReflections theme.

cindyli commented on README.md:
```diff
@@ -15,96 +15,48 @@ This project also uses [Node.js](https://nodejs.org), [Kettle](https://github.co
-- If your new content contains any new wording, you can create new message bundle files or add them to the existing message files (in the `ui/src/messages` directory) and refer to those message names in the handlebars template(s). Please note that a `ui` component specifies and loads only one message bundle at a time.
+The "base" files included in this repository, located in the `themes/base` directory, are intended to provide a bare-bones implementation of the Storytelling Tool. The interface is extensible and customizable to enable the creation of new experiences, themes and story contexts with the tool. To create a custom theme, follow these steps:
+- Set the `theme` value in the server config file (see [Running the site](#Running-the-site)) to the name of your new theme. For the sake of this example, we'll use `cuteCats`.
+- Create a new folder for code and assets associated with the new theme (for examples, please see the `themes/karisma` or `themes/learningReflections` folders)
```
Please also describe the name of the new folder needs to be same as the `theme` value defined in the server config file.

cindyli commented on src/server/serverSetup.js:
```diff
@@ -158,6 +180,18 @@ fluid.defaults("sjrk.storyTelling.server.app.storyTellingHandlers", {
             "route": "/admin/deleteStory/:id",
             "method": "get"
         },
+        testDataHandler: {
```
What is `testDataHandler` for? There isn't a /testData sub-directory in the sjrk root directory.

BlueSlug replied:
There's a testData directory in the /tests/ directory, which contains some files used in both the node and browser tests. The contents are served so the browser tests can refer to them. I didn't want to serve/expose the entire /tests/ directory, but considering we're keeping the node and browser tests separate, I didn't want to keep the testData directory in one or the other of those two locations.

cindyli replied:
Got it. Thanks.

cindyli commented on src/server/serverSetup.js:
```diff
@@ -114,18 +119,35 @@ fluid.defaults("sjrk.storyTelling.server", {
+                    currentTheme: {
+                        type: "kettle.middleware.static",
+                        options: {
+                            root: "@expand:sjrk.storyTelling.server.getCustomThemeFolder({server}.options.globalConfig.theme, ./themes/%theme)",
```
The base theme directory `./themes` can be defined as a configurable component option.

BlueSlug replied:
Do you mean to suggest that I could configure the root directory for the baseTheme middleware? In that case I intentionally chose not to do this, in order to avoid confusion on the part of someone else who would like to implement their own custom theme. I could make it a configurable option and simply not pull that from `sjrk.storyTelling.server.config.json5`, but I thought this was a reasonable implementation, since I only wanted to change it for node tests.

Alternatively, are you suggesting that in this particular case that the second argument for `getCustomThemeFolder()` should be an option instead of hardcoded here?

cindyli replied:
Sorry about the confusion. The "configurable" I meant was to define the base theme dir "./themes" as a component option rather than via `sjrk.storyTelling.server.config.json5` for end users. I think this flexibility will make it easier for reusing or configuring the component as well as writing tests as you said.

cindyli commented on src/server/serverSetup.js:
```diff
@@ -196,3 +229,19 @@ sjrk.storyTelling.server.resolveJSONFile = function (jsonFilePath) {
+        if (fs.existsSync(resolvedFolder)) {
+            folder = resolvedFolder;
+        } else {
+            folder = undefined;
```
When the custom theme folder is not found, starting the server will throw a fatal error. It would be more user friendly to report a proper error message using fluid.fail() and quit.

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -0,0 +1,143 @@
+            storyPromise.reject(error);
+        });;
+    } else {
+        storyPromise.reject();
```
Please provide a proper error message. Also, can we use a common structure for all returned errors:
```
{
    isError: true,
    message: {the-actual-error-message}
}
```
to distinguish errors and non-errors. This structure has already been used in [staticMiddlewareSubdirectoryFilter.js](https://github.com/BlueSlug/sjrk-story-telling/blob/SJRK-237/src/server/middleware/staticMiddlewareSubdirectoryFilter.js#L40).

cindyli replied:
This comment applies to the entire code base. Thanks.

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -0,0 +1,143 @@
+    var callbackFunction = typeof callback === "function" ? callback : fluid.getGlobalValue(callback);
+
+    $.get("/clientConfig").then(function (data) {
+        var theme = themeOverride ? themeOverride : data.clientConfig ? data.clientConfig.theme : "base";
```
The fall-back logic to the base theme should be on the server side as the first thing when global config is read in, which can then be used by the following code as a proper theme. It's better to handle this kind of case as early as possible.

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -0,0 +1,143 @@
+                loadPromise.reject(error);
+            });
+        } else {
+            callbackFunction("page");
```
Why is the parameter "page" for loading the base theme?

cindyli replied:
I see. Using the parameter "page" is for loading the default component used by the base theme such as "sjrk.storyTelling.page.storyBrowse". Can you add a comment to explain? Thanks.

cindyli commented on README.md:
```diff
@@ -15,96 +15,48 @@ This project also uses [Node.js](https://nodejs.org), [Kettle](https://github.co
+    * Docker is an easy way to do this, we recommend the `apache/couchdb` image. To use that image via Docker, run this command: `docker run -p 5984:5984 -d apache/couchdb`
+* Run `node .\src\server\db\dbSetup.js` to configure necessary CouchDB databases (this will also ensure your CouchDB instance is set up in [single-node mode](https://docs.couchdb.org/en/stable/setup/single-node.html))
     * If you have an [admin user](https://docs.couchdb.org/en/stable/intro/security.html) configured on you CouchDB instance, you will need to provide the credentials to `dbSetup.js`. This can be done by setting the `COUCHDB_URL` environment variable. `dbSetup.js` will use the URL specified in `COUCHDB_URL` when connecting to CouchDB. For example: `COUCHDB_URL=http://admin:adminpassword@localhost ...
+* Create a `sjrk.storyTelling.server.config.json5` file in the style of `sjrk.storyTelling.server.config.json5.example`. This file configures the server and allows you to customize the current theme or configure a new one. Edit the values in the `globalConfig` section to do this. More on this in [Theme Customization](#Theme-Customization).
```
It would be helpful to create a section to describe all properties and their purpose, that can specified in `globalConfig`.

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -0,0 +1,143 @@
+    var loadPromise = fluid.promise();
+
+    var cssUrl = fluid.stringTemplate("/css/%theme.css", {theme: theme});
+    var scriptUrl = fluid.stringTemplate("/js/%theme.js", {theme: theme});
```
It seems limited to only allow users to write one js file and one css file for their custom themes. Users would have more freedom if these type of files can be included in their custom templates such as a header template.

Since this approach satisfies the requirement for all two custom themes we have so far, let's stay with allowing only one for each. It can be improved when the requirement goes beyond its limit.

BlueSlug replied:
I agree, and I've filed [SJRK-256](https://issues.fluidproject.org/browse/SJRK-256) to keep track of this suggestion. :)

cindyli commented on tests/server/testServerWithBaseTheme.js:
```diff
@@ -0,0 +1,109 @@
+    }, {
+        event: "{clientConfigRequest}.events.onComplete",
+        listener: "sjrk.storyTelling.server.verifyGetClientConfigSuccessful",
+        args: ["{arguments}.0", "{arguments}.1", undefined]
```
As said earlier, the fall back to the base theme should have already been kicked in here.

cindyli commented on tests/server/testServerWithBaseTheme.js:
```diff
@@ -0,0 +1,109 @@
+    }]
+}];
+
+kettle.test.bootstrapServer(sjrk.storyTelling.server.testServerWithBaseThemeDefs);
```
We should test rendering of actual theme UIs. Is it where you have issues with UIO - https://issues.fluidproject.org/browse/SJRK-255?

BlueSlug replied:
Yes, the tests I was trying to add when I ran into the SJRK-255 issue were to add calls to `sjrk.storyTelling.loadStoryFromParameter()` and `sjrk.storyTelling.loadBrowse()`, which would each load in actual themed UIs.

Mysteriously, while working on SJRK-255 today, I found myself unable to reproduce the error.

BlueSlug committed:
SJRK-237: added some clarifications to the README based on PR review

BlueSlug committed:
SJRK-237: improved error handling and moved "page"/"base" theme logic

cindyli commented on README.md:
```diff
@@ -5,106 +5,66 @@ https://sojustrepairit.org/
+
+| Setting | Description |
+|---------|-------------|
+| `port`  | _(Optional)_ The HTTP port the server will be hosted on. The default value is 8081. |
```
Looking at the source code, `globalConfig` also supports: binaryUploadDirectory, uploadedFilesHandlerPath, deletedFilesRecoveryPath, secrets.adminPass.

Since `sjrk.storyTelling.server.config.json5` is in json5 format, it might be helpful to write a comment section at the start to list all supported values and their goals, as a reminder to end users and developers.

BlueSlug replied:
The three upload-related keys (`binaryUploadDirectory`, `uploadedFilesHandlerPath`, and `deletedFilesRecoveryPath`) were previously pulled in from the configs file, but after further thought I decided these are not "quick and easy" things to configure. If these values are changed, then the code needs to be updated to create and include the referenced directories, as well as to ensure they have the correct permissions set. This means it's non-trivial, and so I think if anyone is going to be digging that far into customizing the application, they'll be able to find the values in `serverSetup.js`. Is that a reasonable assumption?
I suppose I could put comments in the JS to explain their purpose, if that helps.

As for `secrets.adminPass`, this is something I believe we've discussed in the past. The admin password is something that should be kept closely guarded, and having it alongside more "open" values may be misleading or lead to it being overlooked. I suppose I could keep it at the `secrets` path within `globalConfig` to make that more obvious, but I think we decided that having it in a separate file means that the developer will make a conscious move to create that file and keep track of it separately. It's not a perfect solution in either case, but it may prevent accidental disclosure. I'm not 100% sold on this idea haha

cindyli replied:
If these properties are not intended to be configured by end users, they probably should be moved out of `globalConfig` into component options or even member options to prevent them from being accidently overwritten or potential security issues with the present of `adminPass`.

cindyli commented on README.md:
```diff
@@ -5,106 +5,66 @@ https://sojustrepairit.org/
+|---------|-------------|
+| `port`  | _(Optional)_ The HTTP port the server will be hosted on. The default value is 8081. |
+| `savingEnabled` | Specifies whether editing and saving are allowed. If the value is set to `false`, then the site is effectively in "read-only" mode and will behave as a collection of stories rather than an authoring tool. |
+| `theme` | _(Optional)_ The theme to load the site with. If this isn't specified, a base theme will be loaded. More on this in [Theme Customization](#Theme-Customization). |
```
Please also talk about all supported themes: base, karisma, learningReflections.

cindyli commented on README.md:
```diff
@@ -5,106 +5,66 @@ https://sojustrepairit.org/
+| `port`  | _(Optional)_ The HTTP port the server will be hosted on. The default value is 8081. |
+| `savingEnabled` | Specifies whether editing and saving are allowed. If the value is set to `false`, then the site is effectively in "read-only" mode and will behave as a collection of stories rather than an authoring tool. |
+| `theme` | _(Optional)_ The theme to load the site with. If this isn't specified, a base theme will be loaded. More on this in [Theme Customization](#Theme-Customization). |
+| `themeIndexFile` | _(Optional)_ The file to serve at the site root. E.g. `"themeIndexFile": "index2.html"`. The default is `storyBrowse.html` |
```
The statement of "The default is `storyBrowse.html`" seems inaccurate. My understanding by looking at the code is:
1. when the "base" theme is used, the value of "themeIndexFile" will be ignored and be automatically set to storyBrowse.html
2. when a custom theme is used, the value of "themeIndexFile" must be provided.

Please correct me if I'm wrong. If there are such details, better to describe them to avoid confusion. Thanks.

BlueSlug replied:
The case you've described in `1.` is accurate; if `theme` isn't specified, it will load `storyBrowse.html` regardless of whether `themeIndexFile` is specified. I wonder if it makes sense to have `themeIndexFile` control the root file that gets served in all cases.
As for `2.`, the site will work if `theme` is specified and `themeIndexFile` is not. In this case, it will load the `baseTheme` index file (which is `storyBrowse.html`) but with the custom theme applied.

BlueSlug replied:
Would it be better to add a note describing how `themeIndexFile` will be ignored if `theme` is not present and file a Jira for the improvement, or should I try to make it apply regardless of whether a custom theme is loaded?

cindyli replied:
Thanks for the clarification. Yes, it would be better to describe how `themeIndexFile` works as well as the fact that the "base" theme is the fall back theme when `theme` is not specified.

Can you elaborate on "should I try to make it apply regardless of whether a custom theme is loaded"? 

Thanks.

BlueSlug replied:
Sure :) what I meant by "apply regardless of whether a custom theme is loaded" is to implement the `themeIndexFile` (maybe renaming it to simply `indexFile` or `rootFile` or `defaultDocument`) in such a way that it specifies the file to serve at the site root regardless of what theme is used. I.e. if the developer wants to use `base` but change the default file from `storyBrowse.html` to, say, `storyEdit.html`, they can use the same key in `globalConfig` instead of altering `serverSetup.js`

cindyli replied:
This change doesn't seem necessary at least for now. Let's stick with `themeIndexFile` and have a clear document for it. Thanks.

cindyli commented:
UIO is broken with all three themes. The console error shows `TypeError: text is undefined`.

<img width="598" alt="Screen Shot 2019-07-19 at 10 42 44 AM" src="https://user-images.githubusercontent.com/572877/61543774-1fbb1b80-aa12-11e9-893b-e1380a98e015.png">

BlueSlug commented:
> UIO is broken with all three themes. The console error shows `TypeError: text is undefined`.

I'm having trouble reproducing this issue, and though UIO is not loading for me either, the error message I'm seeing is consistent with the one I reported in SJRK-255:
```
fluidParser.js:420 Uncaught TypeError: Cannot read property 'replace' of undefined
    at Object.fluid.XMLEncode (fluidParser.js:420)
    at renderComponent (fluidRenderer.js:980)
    at renderComponentSystem (fluidRenderer.js:1139)
    at renderRecurse (fluidRenderer.js:1327)
    at Object.that.renderTemplates (fluidRenderer.js:1405)
    at Object.fluid.reRender (fluidRenderer.js:1488)
    at Object.fluid.render (fluidRenderer.js:1555)
    at Object.that.render (RendererUtilities.js:104)
    at fluid.rendererComponent.refreshView (RendererUtilities.js:178)
    at invokeInvoker (FluidIoC.js:1749)
```

cindyli commented:
I encountered the same error you had when using Chrome. The error `TypeError: text is undefined` is with Firefox.

BlueSlug commented:
I was just testing in Firefox when you replied :) Okay, well in either case it seems to be related to UIO attempting to render before its localized messages have been fully loaded, which I intended to tackle in SJRK-255. How should I proceed?

cindyli commented:
I feel we should tackle it as part of this reorganization work. It won't be pretty if the live SJRK site has a broken UIO.

BlueSlug committed:
SJRK-237: documentation and function updates according to PR review

BlueSlug committed:
SJRK-237: fixed UIO bug and added tests for dynamic themed page loading
- UIO is still erroring out occasionally in the test page, but on the 
site proper it's working 99% of the time or more

BlueSlug committed:
SJRK-237: commented out potentially-failing tests. Sort out in SJRK-255.

BlueSlug committed:
SJRK-237: moved base-themed images to base theme folder, removed unused

BlueSlug committed:
SJRK-237: moved clientConfig theme fallback logic to server side

BlueSlug committed:
SJRK-237: parameterized themes directory for ease of config, decoupling

cindyli commented on Gruntfile.js:
```diff
@@ -16,10 +16,10 @@ module.exports = function (grunt) {
         },
         jsonlint: {
-            all: ["package.json", ".eslintrc.json", "server/src/**/*.json", "ui/src/**/*.json", "tests/**/*.json"]
+            all: ["package.json", ".eslintrc.json", "themes/**/*.json", "tests/**/*.json"]
```
Json5 files such as "sjrk.storyTelling.server.config.json5" and those in the "tests/" directory should be linted too.

BlueSlug replied:
jsonlint does not currently support json5 format, nor am I able able to find a viable eslint plugin for json5. That said, if I remove the comments from the json5 files, they should pass jsonlint, but if any developer decides to use any of the extra features of json5 they will fail the linting checks.

cindyli replied:
The universal repo has [an example](https://github.com/GPII/universal/blob/master/Gruntfile.js#L23) of linting json5 files.

amb26 replied:
The relevant module is https://github.com/GPII/gpii-grunt-lint-all , written by @the-t-in-rtf which is well worth using

cindyli commented on README.md:
```diff
@@ -5,106 +5,73 @@ https://sojustrepairit.org/
+- Create a new folder for code and assets associated with the new theme (for examples, please see the `themes/karisma` or `themes/learningReflections` folders). The folder name **must** be the same as the `theme` value in the config file.
+- Create a JavaScript file which contains extensions of the `page` grades (e.g. for `sjrk.storyTelling.page.storyEdit`, add a new grade `sjrk.storyTelling.cuteCats.storyEdit` which has the former as one of its gradeNames). These extensions could include new `ui` components for new sections of the page, new events or other functionality. The name of the JavaScript file must match the folder and th ...
+- Create a CSS file in the `css` directory with all of the styling rules specific to the new theme. The CSS file, like the JavaScript file, should have the same name as the theme and theme folder: `themes/cuteCats/css/cuteCats.css`
+- Add any new associated [handlebars](https://handlebarsjs.com/) templates to be used by new UI components
```
Indicate these templates need to be in the `templates` directory.

cindyli commented on sjrk.storyTelling.server.config.json5.example:
```diff
@@ -0,0 +1,25 @@
+            "server": {
+                "options": {
+                    "globalConfig": {
+                        // the HTTP port the server will hosted from, default is 8081
```
will -> will be

cindyli commented on sjrk.storyTelling.server.config.json5.example:
```diff
@@ -0,0 +1,25 @@
+                        // a default file to load at the site root (e.g. index.html)
+                        "themeIndexFile": "introduction.html",
+
+                        // if true, saving stories to the database will be enabled
```
As you said in the readme, this flag is actually to enable/disable the authoring part. It's probably more accurate to call it "authoringEnabled".

BlueSlug replied:
I think the comment in the example json5 file is actually more accurate, and the comment in the readme is descriptive of the behaviour of _both_ the server flag and the UI-side flag, which are still separately defined. I will add a comment in the readme to indicate that both should be set together and should match each other.

cindyli replied:
Why does UI side need its own flag? /clientConfig is returning "savingsEnabled" value that is read from the server side config file.

BlueSlug commented:
Previously identified, but a good time to tackle it: serverSetup.js dataSource dist opts duplication - make it more DRY by either adding it to the base grade or the parent component

cindyli commented on src/server/serverSetup.js:
```diff
@@ -168,31 +205,53 @@ fluid.defaults("sjrk.storyTelling.server.app.storyTellingHandlers", {
+        themePath = fluid.stringTemplate(themeFolder, { theme: theme });
+
+        if (!fs.existsSync(themePath)) {
+            fluid.fail("The custom theme folder " + themePath + " does not exist. Please verify that the theme name is configured properly.");
```
The failure reported will include your message as well as the detailed tech error by printing out the component tree. The way I found to report a graceful error and exit is to replace this line with:

```
console.log("FATAL ERROR: ...");
process.exit(1);
```

BlueSlug replied:
I implemented this suggestion, and while it did kill the server gracefully, it was not something I was able to test against to ensure that this was what was happening. Since the `process.exit(1)` call simply kills the server, I would have to use or write some sort of external tool to test the output in the console.

The detailed component tree printout is verbose and would make finding the actual error message difficult, but I think there's value in doing things the "Infusion way" and calling fluid.fail. I wonder if there is a way to call fluid.fail without having it print out the component tree, but just the message being passed into it.

@amb26 do you have any thoughts on this matter?

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -0,0 +1,175 @@
+                });
+            });
+        } else {
+            theme = "base";
```
This line is unnecessary. All exceptions such as undefined theme should have been handled at the server side. This "if...else..." block can simply check `theme === "base"` or else.

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -0,0 +1,175 @@
+            loadPromise.resolve(theme);
+        }
+    }, function (jqXHR, textStatus, errorThrown) {
+        callbackFunction("base"); //default to the base theme page
```
replace "base" with the variable `theme` in case there's a name change for the base theme in the future. Please avoid hardcoded strings whenever possible. Thanks.

BlueSlug replied:
In this case the request to `/clientConfig` on the server will have failed, so `theme` will be undefined (it's out of scope in either case). Would it be acceptable to pass in something like `baseThemeName`  in the call to `loadThemedPage()` and use that here? Likewise, I could make that an option in the `secureConfig` collection and refer to that whenever it needs to be used in order to avoid hardcoded strings.

cindyli replied:
If the server doesn't work, would the client be able to operate on its own? If not, it might be more helpful to report an error rather than carrying on the operation.

cindyli commented on tests/server/configs/sjrk.storyTelling.server.testServerWithBaseTheme.json5:
```diff
@@ -0,0 +1,29 @@
+            "server": {
+                "options": {
+                    "globalConfig": {
+                        "binaryUploadDirectory": "./tests/server/uploads",
```
Apparently tests don't make use of config values at line 9-10. Otherwise they won't pass with these configs being moved to `secureConfig`.

It would be nice to have integration tests to cover the communication between server and client so these configs make sense in tests.

cindyli commented on tests/server/configs/sjrk.storyTelling.server.testServerWithBaseTheme.json5:
```diff
@@ -0,0 +1,29 @@
+                        "uploadedFilesHandlerPath": "./tests/server/uploads"
+                    },
+                    "components": {
+                        "storyDataSource": {
```
This subcomponent doesn't seem to be used in the test too. It would be less confusing to only define the necessary bits that have been used by the test.

This comment applies to other config files too. Thanks.

cindyli commented on tests/ui/js/storyTellingServerUITests.js:
```diff
@@ -0,0 +1,173 @@
+        });
+    });
+
+    // var browsePageOptions = {
```
Are these commented lines are for the test meant to be dealt with at https://issues.fluidproject.org/browse/SJRK-255?

BlueSlug replied:
That's correct, and the same is true of the last two elements of the test sequence

cindyli replied:
Can you track these tests-in-progress in a separate branch so we can merge clean code into the master. Thanks.

cindyli replied:
bump

cindyli replied:
bump again. Please remove the commented code and track it in your own branch so we can merge the clean code. Thanks.

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -0,0 +1,175 @@
+ *               regardless of whether theme information was specified or retrieved
+ * - "themeOverride": allows overriding of the theme stored in the configuration
+ */
+sjrk.storyTelling.loadThemedPage = function (callback, themeOverride) {
```
It's weird to accept "themeOverride" parameter that could override the theme sent back from the server, which is the theme defined in the config file. As what's shown in your test case, what would happen if this value is provided as `null` or `undefined`? Does the client side have to perform its own fall back to the base? 

Could you elaborate the purpose of "themeOverride" parameter? Thanks.

BlueSlug replied:
I had added that in order to do more robust testing of the `loadThemedPage` function, since it's a browser test and not based on varying server config files. This is the only usage of it that parameter that I've anticipated. If `themeOverride` is falsy, it will fall back to the theme returned from the request to `/clientConfig`, according to line 118.

I suppose the structure of this function should be reworked, if `themeOverride` is going to be kept, since some of the possible cases are not handled, namely if the request fails but `themeOverride` is truthy, or if the request fails and `themeOverride` is truthy but also not valid.

cindyli replied:
A better way to test it is to create a mocking server response for /clientConfig, which mocks http responses you are expecting for your test cases. Infusion uses "Sinon.js" library that also has [the feature of creating a fake server](https://sinonjs.org/#fake-server).

BlueSlug committed:
SJRK-237: clarified documentation

BlueSlug committed:
SJRK-237: added tests for configuration with missing theme

BlueSlug committed:
SJRK-237: removing unused bits of code

BlueSlug commented:
> Previously identified, but a good time to tackle it: serverSetup.js dataSource dist opts duplication - make it more DRY by either adding it to the base grade or the parent component

I was not able to achieve this. In all cases, the server would fail to load.

cindyli committed:
GPII-237: Simplify the distribution of the host value to data sources.

cindyli commented:
> > Previously identified, but a good time to tackle it: serverSetup.js dataSource dist opts duplication - make it more DRY by either adding it to the base grade or the parent component
> 
> I was not able to achieve this. In all cases, the server would fail to load.

I issued [this pull request](https://github.com/BlueSlug/sjrk-story-telling/pull/25) to your SJRK-237 branch to address the issue.

BlueSlug committed:
SJRK-237: added linting for json5 files, using gpii-grunt-lint-all now

BlueSlug committed:
Merge pull request #25 from cindyli/SJRK-237

SJRK-237: Simplify the distribution of the host value to data sources.

BlueSlug committed:
Merge branch 'SJRK-237' of https://github.com/BlueSlug/sjrk-story-telling into SJRK-237

BlueSlug commented:
> > > Previously identified, but a good time to tackle it: serverSetup.js dataSource dist opts duplication - make it more DRY by either adding it to the base grade or the parent component
> > 
> > 
> > I was not able to achieve this. In all cases, the server would fail to load.
> 
> I issued [this pull request](https://github.com/BlueSlug/sjrk-story-telling/pull/25) to your SJRK-237 branch to address the issue.

That was the one case I had neglected to try :0 I'd tried things that mostly involved incorporating the expander into the base grade and not the parent component.

Thanks @cindyli!

BlueSlug committed:
SJRK-237: added sinon for mocking server responses in UI-to-server tests

BlueSlug committed:
SJRK-237: removed hardcoded values from UI-server code, updated tests

BlueSlug committed:
SJRK-237: updated `page` grade names to make switching more convenient

BlueSlug committed:
SJRK-237: renamed page files to reflect renaming, fixed creator calls

BlueSlug commented:
> What do you think about calling it "base" rather than "baseTheme"?

I thought about this a little bit before arriving at `baseTheme`, and the main reason is that we have other parts of the codebase (heh) where we use the term "base" to indicate a grade that is used primarily as the common source of multiple other grades. While the purpose of the base theme is similar, it also stands alone, so I thought it would be better to be explicit in naming it after a theme. This also helps indicate where the theme name would go for any new themes (this is reflected in the readme)

BlueSlug committed:
SJRK-237: renamed ".base" grades to ".core" to differentiate from theme

BlueSlug committed:
SJRK-237: savingEnabled -> authoringEnabled

BlueSlug committed:
SJRK-237: baseTheme -> base, to be consistent with the theme folder name

BlueSlug commented:
@cindyli ready for further review. In my testing I discovered an issue with the empty block filter and captured it as [SJRK-263](https://issues.fluidproject.org/browse/SJRK-263)

cindyli commented on README.md:
```diff
@@ -35,7 +35,7 @@ While Infusion allows for just about any possible extension you can imagine and
 |---------|-------------|
 | `port`  | _(Optional)_ The HTTP port the server will be hosted on. The default value is 8081. If this is changed, it should also be updated in `Dockerfile`, `docker-compose.dev.yml` and `docker-compose.cloud.yml` |
-| `savingEnabled` | Specifies whether saving stories to the database is allowed. If the value is set to `false` and the matching option in `sjrk.storyTelling.baseTheme.page`, then the site is effectively in "read-only" mode and will behave as a collection of stories rather than an authoring tool. For best results, the values in this config file and the `page` grade should match each other. |
+| `authoringEnabled` | Specifies whether creating and saving stories to the database is allowed. If the value is set to `false` and the matching option in `sjrk.storyTelling.base.page`, then the site is effectively in "read-only" mode and will behave as a collection of stories rather than an authoring tool. For best results, the values all `page` grades should be derived from this config file (th ...
```
I think we've decided it's no longer necessary to define the client side `authoringEnabled` flag in `sjrk.storyTelling.base.page`. The client will now fetch this flag from the server.

BlueSlug replied:
The value that is ultimately used is the one passed into from the server, but there are still some bits of the templates that are controlled by that flag, as well as at least one IoC reference to it in `storyEdit`, so I've kept it in the `page` grade intentionally. I also intentionally left the reference in the documentation, though it makes sense to remove it from there at this point.

cindyli replied:
Why cannot them use the `authoringEnabled` flag passed into from the server as all page components are instantiated with this server flag passed in in their options?

cindyli commented on package.json:
```diff
@@ -1,41 +1,43 @@
+    },
+    "devDependencies": {
+        "chance": "1.0.16",
+        "eslint-config-fluid": "1.1.0",
```
With the use of "gpii-grunt-lint-all", please check if all other dependencies for linting can be removed, such as "eslint-config-fluid", "fluid-grunt-eslint", "grunt-eslint", "grunt-jsonlint". Thanks.

cindyli commented on Gruntfile.js:
```diff
@@ -15,20 +15,19 @@ module.exports = function (grunt) {
-            all: ["package.json", ".eslintrc.json", "server/src/**/*.json", "ui/src/**/*.json", "tests/**/*.json"]
+        lintAll: {
+            sources: {
+                js: ["src/**/*.js", "tests/**/*.js", "*.js"],
```
There are also js flies in "themes/" directory. Please add. Thanks.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -338,6 +348,31 @@ sjrk.storyTelling.server.deleteSingleFileRecoverable = function (fileToDelete, d
+sjrk.storyTelling.server.getClientConfig = function (request, globalConfig, secureConfig) {
+    var clientConfig = {
+        theme: globalConfig.theme || secureConfig.baseThemeName,
+        baseTheme: secureConfig.baseThemeName,
```
As we discussed last Friday, as the client no longer needs to distinguish the base theme from other themes with the change of embedding "base" in component names, the server doesn't need to return `baseTheme: "base"` to the client.

BlueSlug replied:
I kept it in there for the check in `sjrk.storyTelling.loadThemedPage`, since we don't want to load `base.js` or `base.css` (which don't exist) and we need something to compare against in order to make that evaluation.

cindyli replied:
We should discuss if a header template should be added to handle the inclusion of css and js from theme directories which should also help to solve [SJRK-256 issue](https://issues.fluidproject.org/projects/SJRK/issues/SJRK-256).

BlueSlug replied:
That sounds like a good idea, though I feel it might be a bit out of scope for this Jira. I agree that we should talk about it for sure :)

cindyli replied:
I agree this can be addressed in a separate jira. Can you create one and link with SJRK-256? Thanks.

BlueSlug replied:
[SJRK-264](https://issues.fluidproject.org/browse/SJRK-264) has been filed.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -338,6 +348,31 @@ sjrk.storyTelling.server.deleteSingleFileRecoverable = function (fileToDelete, d
+    };
+
+    request.events.onSuccess.fire({
+        message: "Client configuration request successfully received",
```
Indicating `isError: false` for rejected requests should be enough for the client to distinguish success and rejected response. Returning `clientConfig` object should be good enough: `request.events.onSuccess.fire(clientConfig)`.

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -0,0 +1,174 @@
+/*
```
It's confusing to have both "server" and "ui" in the file name. Would it be better to rename it to "storyTellingUiHandler" or "storyTellingUiFetcher"?

It also looks redundant to have almost all files in src/ui/ directory starting with "storyTelling" which make file names longer. The project name has been clear about "story-telling".

BlueSlug replied:
I agree on all counts. The files in the `ui` directory were named roughly according to their grade names in order to help make finding the relevant files easier, though this naming convention is out of keeping with the files in the `server` directory. This is, more than anything else, simply a remnant of the way we kicked off the project with a few days of quick-and-dirty development and a fair amount of premature optimization. I will remove "storyTelling." from all file names and think of a more appropriate name for `storyTellingServerUI.js`.

BlueSlug replied:
In a related issue, I think it would make sense at some point to review all grade names and make sure they still make sense, given that the project has changed so much in the last two years

BlueSlug replied:
Re `storyTellingServerUI.js`, I think it might need more than just a renaming. It might make more sense to refactor these functions a bit and put them in other files. `loadBrowse` and `loadStoryFromParameter`, in addition to possibly being renamed themselves, would more logically live in the same file as the `storyBrowse` and `storyView` grades they end up calling. Also my usage of promises and callback parameters simultaneously isn't ideal, so this might be a good opportunity to switch to an all-promises setup.

cindyli replied:
What a coincident that I've been thinking about refactoring of this file too. I actually think all functions in this file can be moved into components and all workflows can be incorporated and controlled by individual components. Use "Browse story" as an example:

1. move the fetch of "/clientConfig" into "sjrk.storyTelling.base.page" `onCreate` event;
2. The retrieval of stories to "/stories" endpoint can be moved to  "sjrk.storyTelling.base.page.storyBrowse" component, in parallel with the fetch of "/clientConfig". It also fires an event "onStoriesReady" when stories are received.
3. The aggregate event for`onClientConfigReady + onStoriesReady` controls other parts such as the load of custom theme files, the instantiation of templateManager and the page rendering.

Considering our timeline, it might be better to do this improvement in a separate JIRA if we decide to do it.

BlueSlug replied:
[SJRK-265](https://issues.fluidproject.org/browse/SJRK-265) filed to address this

cindyli commented on tests/server/configs/sjrk.storyTelling.server.testServerWithBaseTheme.json5:
```diff
@@ -0,0 +1,19 @@
+                    "components": {
+                        "baseTheme": {
+                            options: {
+                                root: "./tests/testData/mockBaseTheme"
```
Rather than directly overriding `root`, its calculation can be derived by defining configs via globalConfig and secureConfig, which will help to test sjrk.storyTelling.server.getThemePath() function too.

The same comment applies to definitions in tests/server/configs/sjrk.storyTelling.server.testServerWithCustomTheme.json5.

cindyli commented on tests/server/testServerWithMissingTheme.js:
```diff
@@ -0,0 +1,35 @@
+            configName: "sjrk.storyTelling.server.testServerWithMissingTheme",
+            configPath: "./tests/server/configs"
+        });
+    }, "The custom theme folder");
```
Can you please verify a more meaningful error message? Thanks.

cindyli commented:
Running `grunt lint` reports this warning:

```
Running "wrapped-eslint:md" (wrapped-eslint) task
Could not find any files to validate.
```

BlueSlug committed:
SJRK-237: removed old linter dependencies, updated docs

BlueSlug committed:
SJRK-237: renamed all UI js files to remove repetitive `storyTelling.`

BlueSlug committed:
SJRK-237: streamlined clientConfig return, updated mock fixture to match

BlueSlug committed:
SJRK-237: updated server theme tests to use secureConfig options, tidier

BlueSlug committed:
SJRK-237: added Markdown linting and fixed all Markdown linting issues

BlueSlug commented:
@cindyli ready for review :)

cindyli commented on README.md:
```diff
@@ -1,140 +1,176 @@
+story contexts within the tool.
+The site comes with two custom themes already available: `karisma` and `learningReflections`.
+
+* `learningReflections` Has a few extra pages with content and is styled to illustrate the support of the Storytelling
```
Has -> has

cindyli commented on src/ui/base-page.js:
```diff
@@ -11,19 +11,19 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
         pageSetup: {
             resourcePrefix: "",
-            savingEnabled: true // should match the setting in sjrk.storyTelling.server.server.globalConfig
+            authoringEnabled: true // should match the setting in sjrk.storyTelling.server.config.json5
```
The current comment is wrong. Please move this line into a commented section and explain it must be provided by integrators and what it is for.

BlueSlug committed:
SJRK-237: fixed readme typo

BlueSlug committed:
SJRK-237: updated comment for authoringEnabled flag

BlueSlug commented:
@cindyli ready for review

BlueSlug committed:
SJRK-237: removed currently-unused code
